Fixing guidelines for the scheduler.

indualagarsamy opened the pull request:
Relates to: https://github.com/Particular/Requirements/issues/188
Clarified the limitations of the scheduler. 

**Questions:**
- Do you think its better to separate out the last section to its own article?

indualagarsamy commented:
@Particular/documentation - can you review what i currently have?

mauroservienti commented on nservicebus/scheduling-with-nservicebus.md:
```diff
@@ -5,19 +5,19 @@ tags:
+## How the scheduler works
 
-In the scenario where you are "polling" for some change in information state you should instead consider a "push" approach.
+The scheduler holds a all list of tasks scheduled in a non-durable in-memory dictionary. In version 3 and 4 tasks are scoped per `AppDomain`. In version 5 they are scoped per Bus instance.
```
"holds a all list" should be "holds a list"

indualagarsamy replied:
:) Will fix it.

mauroservienti commented on nservicebus/scheduling-with-nservicebus.md:
```diff
@@ -5,19 +5,19 @@ tags:
+The scheduler holds a all list of tasks scheduled in a non-durable in-memory dictionary. In version 3 and 4 tasks are scoped per `AppDomain`. In version 5 they are scoped per Bus instance.
 
-The scheduling API is often used to poll state, for example the file system or an external webservice. In these scenarios it is much more efficient to adopt a push based model. In the case of the file system you should instead consider a [FileSystemWatcher](http://msdn.microsoft.com/en-us/library/system.io.filesystemwatcher.aspx). In the case a webservice should instead consider a [WebHook](http: ...
+When a new scheduled task is created it is given an unique identifier and stored in the endpoint's in-memory dictionary. The identifier for the task is sent in a message to the Timeout Manager for the message to be deferred for the specified time interval. When the specified time elapses, the Timeouts dispatcher returns the message containing the identifier to the endpoint with the scheduled task ...
```
is the Saxon genitive form correct? "endpoint's in-memory" shouldn't is simply be "endpoint in-memory"

indualagarsamy replied:
the in-memory dictionary belongs to the endpoint.

mauroservienti commented on nservicebus/scheduling-with-nservicebus.md:
```diff
@@ -25,25 +25,20 @@ The difference between these examples is that in the latter a name is given for
+- Since the scheduler is non-durable, if the process restarts, all scheduled tasks (that are created during the endpoint's startup) are recreated and given new identifiers. Tasks scheduled before the restart that arrive at the endpoint queue will not be found and a message is written to the log as information.  
+- Already created schedule tasks cannot be canceled or changed.
+- Does not support a task to be started at a specified specific time.
 - Each scheduled task executes on a new thread using the `Task.Factory.StartNew(Action)` method, which means that there will be no transaction scope by default and it is up to you to create one if needed;
```
As far as I know we cannot assume that `Task.Factory.StartNew(Action)` will create a new thread, I'd rephrase it saying that each scheduled task will be executed in a new `Task` generated by the Task Parallel Library using `Task.Factory.StartNew(Action)`...

indualagarsamy replied:
@mauroservienti - will change it per your suggestion. I originally kept the statement as is (from the original doco).

mauroservienti commented on nservicebus/scheduling-with-nservicebus.md:
```diff
@@ -25,25 +25,20 @@ The difference between these examples is that in the latter a name is given for
+- Does not support a task to be started at a specified specific time.
 - Each scheduled task executes on a new thread using the `Task.Factory.StartNew(Action)` method, which means that there will be no transaction scope by default and it is up to you to create one if needed;
-- You will probably only do a `Bus.Send()` or `Bus.SendLocal()` in the task. The handler of the sent message will be invoked as all other handlers and if configured and supported by underlying transport will be wrapped in a transaction as usual;
+- Since the scheduler uses the queuing mechanism, it does have some side effects on the timeliness of scheduled tasks. When a task is scheduled to be run at a given time it is actual not "executed at that time", it is instead "queued at that time" to be executed. In most cases this distinction will have no noticeable effect on the behavior of the the scheduling API. However in high load systems t ...
```
on the behavior of **the the** scheduling API

mauroservienti commented on nservicebus/scheduling-with-nservicebus.md:
```diff
@@ -25,25 +25,20 @@ The difference between these examples is that in the latter a name is given for
 - Each scheduled task executes on a new thread using the `Task.Factory.StartNew(Action)` method, which means that there will be no transaction scope by default and it is up to you to create one if needed;
-- You will probably only do a `Bus.Send()` or `Bus.SendLocal()` in the task. The handler of the sent message will be invoked as all other handlers and if configured and supported by underlying transport will be wrapped in a transaction as usual;
+- Since the scheduler uses the queuing mechanism, it does have some side effects on the timeliness of scheduled tasks. When a task is scheduled to be run at a given time it is actual not "executed at that time", it is instead "queued at that time" to be executed. In most cases this distinction will have no noticeable effect on the behavior of the the scheduling API. However in high load systems t ...
+- The scheduler API does not support scaling out the endpoint on the same machine or doing a side-by-side deployment of an endpoint. In both these cases, you have multiple endpoint instances sharing the same queue where the schedule task arrives.  As the scheduler is not-durable and the task details are stored in the endpoint instance that created the task, the task may or may not be found/execut ...
```
I understood the above, but I'd say that it is not clear enough, the reader needs to connect the dots understanding that each time a new Task is scheduled a **new** task id is created, thus 2 instances of the same endpoint will create the same task with different ids, then the issue. Correct?

indualagarsamy replied:
I'll clarify it better. The issue is detailed in 188 (requirements)

mauroservienti commented on nservicebus/scheduling-with-nservicebus.md:
```diff
@@ -25,25 +25,20 @@ The difference between these examples is that in the latter a name is given for
 
-## When not to use it
+## Converting a schedule task into a saga
+To store these tasks durably in order to support either a scale-out scenario or a side-by-side deployment scenario, you can use the saga instead. You can think of the scheduler as a never ending saga.
```
use **a** saga instead

SeanFeldman commented on Snippets/Snippets_4/Scheduling/Scheduling.cs:
```diff
@@ -14,7 +14,7 @@ public void ScheduleTask()
         // Name a schedule task and invoke it every 5 minutes
         Schedule.Every(TimeSpan.FromMinutes(5))
-            .Action("Task name", SomeCustomMethod);
+            .Action("MyCustomTask", SomeCustomMethod);
```
Any particular reason for this rename? I'm just curious. Thank you.

indualagarsamy replied:
I wanted the saga name to correlate with the task name. In the last section, I'll make this more obvious. Thanks.

indualagarsamy commented on nservicebus/scheduling-with-nservicebus.md:
```diff
@@ -25,25 +25,26 @@ The difference between these examples is that in the latter a name is given for
-
-## When not to use it
+Using the same example usage of the scheduler API shown above, in order to convert it to a saga:
+1. Create a saga for the specified task name that is started by a message. Create a new message if necessary. In this example, the message that starts the saga is `StartMyCustomTaskSaga`. This message will be sent by the endpoint on startup, using a ` Bus.SendLocal(new StartMyCustomTaskSaga{TaskName = "StartupTask1"});`. If you did not use the API with the task name, name your saga appropriately. ...
```
Yes, will do.

indualagarsamy commented on nservicebus/scheduling-with-nservicebus.md:
```diff
@@ -25,25 +25,26 @@ The difference between these examples is that in the latter a name is given for
+- While the task repeat interval can be specified, does not support a task to be started at a specific time.
+- Each scheduled task will be executed in a new Task generated by the Task Parallel Library using `Task.Factory.StartNew(Action)`, which means that there will be no transaction scope by default and it is up to you to create one if needed;
+- Since the scheduler uses the queuing mechanism, it does have some side effects on the timeliness of scheduled tasks. When a task is scheduled to be run at a given time it is actual not "executed at that time", it is instead "queued at that time" to be executed. In most cases this distinction will have no noticeable effect on the behavior of the scheduling API. However in high load systems the f ...
+- The scheduler API does not support scaling out the endpoint on the same machine or doing a side-by-side deployment of an endpoint. When you have multiple instances of the endpoint that's running on the same machine as is the case with these scenarios, these endpoint instances share the same input queue. Since each endpoint maintains its own created tasks in memory, when the specified time is up ...
```
Yeah, true. Will make it bold and a warning.

indualagarsamy commented:
@mauroservienti @SeanFeldman - I have incorporated your comments. Also rebased with Master, so, should be able to merge this PR. Can you see if all your concerns are answered?
@Particular/documentation - Do you think we should split the saga example to a different article or leave it in this one? Let me know if this good, so I can get this merged soon.

mauroservienti commented:
@indualagarsamy looks good.
@danielmarbach can you add more details on your suggestion?

danielmarbach commented:
@mauroservienti which one?

mauroservienti commented:
@danielmarbach sorry, this:

> Maybe also show how a saga which actually allow to cancel a scheduled task?

andreasohlund commented on Snippets/Snippets_3/Scheduling/SchedulerSaga.cs:
```diff
@@ -0,0 +1,66 @@
+
+    #region ScheduleTaskSaga
+
+    class MyCustomTaskSaga : Saga<MyCustomTaskSagaData>,
```
Would there be better to make this a full sample instead?

danielmarbach commented:
A cancel message which is sent to the saga and the saga is completed.

On Mon, Feb 23, 2015 at 2:44 PM, Mauro Servienti <notifications@github.com>
wrote:

> @danielmarbach <https://github.com/danielmarbach> sorry, this:
>
> Maybe also show how a saga which actually allow to cancel a scheduled task?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/docs.particular.net/pull/571#issuecomment-75542476>
> .
>



-- 

Daniel Marbach

Solution Architect, Particular Software

Email:  daniel.marbach@particular.net

Web:    www.particular.net
<http://t.signauxneuf.com/e1t/c/5/f18dQhb0S7lC8dDMPbW2n0x6l2B9nMJN7t5XX4RzDjlVS9QB88pTc_dW3Ljg7x56dWyBdgBXs402?t=http%3A%2F%2Fwww.particular.net%2F&si=4896787407044608&pi=33cc6202-1f31-4040-8d6c-e70641f44bba>

mauroservienti commented:
@danielmarbach  as @andreasohlund is pointing out what about setting up a full sample?

danielmarbach commented:
As an addition. Yeah great idea. Do you have time? I could do it tomorrow
or after tomorrow. Still have a few things to run today.

mauroservienti commented:
not today, I'm trying to catch up with everything I lost in the last 4 days.

On Mon, Feb 23, 2015 at 2:53 PM, Daniel Marbach <notifications@github.com>
wrote:

> As an addition. Yeah great idea. Do you have time? I could do it tomorrow
> or after tomorrow. Still have a few things to run today.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/docs.particular.net/pull/571#issuecomment-75543792>
> .
>



-- 
Mauro Servienti | Solution Architect | Particular Software

indualagarsamy commented:
@danielmarbach - I've updated the documentation to include the cancel task as a note to the Saga and moved around the `When Not to Use` to be more closer to the `Current Limitations` and also added another point in the When Not to use section regarding feature limitations.
